### PR TITLE
fix(opensearch): handle index not present errors in search api

### PR DESCRIPTION
**Pull request by tsdk02**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- Handled errors when index is not present, by logging the corresponding error message and treating the hits as 0.
- Changed the structure of `OpensearchOutput` and corresponding structs which are used to handle the error.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

The msearch API errors out since the index is not present, so handled the error.

## How did you test it?

- Set up the global search locally
- Test the global search with any keyword
- Go through the log, and `5xx errors` should not be present anymore, with the errors logged, and handling of the index not found errors.
- 
Screenshot 2024-06-12 at 2 40 43 PM
Screenshot 2024-06-12 at 2 41 12 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/analytics/Cargo.toml [hunk 1]
```diff
@@ -53,3 +53,4 @@ strum = { version = "0.26.2", features = ["derive"] }
 thiserror = "1.0.58"
 time = { version = "0.3.35", features = ["serde", "serde-well-known", "std"] }
 tokio = { version = "1.37.0", features = ["macros", "rt-multi-thread"] }
+log = "0.4"
```
Comment (lsampras) [inline crates/analytics/Cargo.toml:56]: we use tracing crate for logging

File: crates/analytics/src/search.rs [hunk 2]
```diff
@@ -26,23 +26,47 @@ pub async fn msearch_results(
         .execute(query_builder)
         .await
         .change_context(OpenSearchError::ConnectionError)?
-        .json::<OpenMsearchOutput<Value>>()
+        .json::<OpenMsearchOutput>()
```
Comment (lsampras) [inline crates/analytics/src/search.rs:29]: Can we read this as a string first using .text() and then try to parse it?
that way we can print the string as part of the parsing error helping us debug this in production environments

File: crates/analytics/src/search.rs [hunk 3]
```diff
@@ -26,23 +26,47 @@ pub async fn msearch_results(
...
+            OpensearchOutput::Success(success) => {
+                if success.status == 200 {
+                    GetSearchResponse {
+                        count: success.hits.total.value,
+                        index,
+                        hits: success
+                            .hits
+                            .hits
+                            .into_iter()
+                            .map(|hit| hit._source)
```
Comment (lsampras) [inline crates/analytics/src/search.rs:47]:
```suggestion
                            .map(|hit| hit.source)
```

The `_` prefix is a rust convention for unused variables can we remove it?

File: crates/analytics/src/search.rs [hunk 4]
```diff
@@ -69,18 +93,42 @@ pub async fn search_results(
...
+                    index: req.index,
+                    hits: success
+                        .hits
+                        .hits
+                        .into_iter()
+                        .map(|hit| hit._source)
+                        .collect(),
+                })
+            } else {
+                error!("Unexpected status code: {}", success.status);
```
Comment (lsampras) [inline crates/analytics/src/search.rs:114]: use the tracing library

**Review by lsampras — COMMENTED** [review]
Can you also update the readme in `crates/analytics/docs` to add conditions for enabling the global search on dashboard as well

I believe we have only mentioned audit_log & system_metrics

File: crates/analytics/src/search.rs (hunk 2 above)
Comment (tsdk02) [reply crates/analytics/src/search.rs]: Done

File: crates/analytics/src/search.rs (hunk 3 above)
Comment (tsdk02) [reply crates/analytics/src/search.rs]: Removed it

File: crates/analytics/src/search.rs (hunk 4 above)
Comment (tsdk02) [reply crates/analytics/src/search.rs]: Done

File: crates/analytics/Cargo.toml (hunk 1 above)
Comment (tsdk02) [reply crates/analytics/Cargo.toml]: Done

**Comment by tsdk02** [comment]
> Can you also update the readme in `crates/analytics/docs` to add conditions for enabling the global search on dashboard as well
> 
> I believe we have only mentioned audit_log & system_metrics

Added `global_search=true` in the README.md file

File: crates/analytics/src/search.rs [hunk 5]
```diff
@@ -22,14 +22,26 @@ pub async fn msearch_results(
...
+    let response_body = match serde_json::from_str::<OpenMsearchOutput>(&response_text) {
+        Ok(parsed_response) => parsed_response,
+        Err(parse_error) => {
+            tracing::error!(
+                "Failed to parse response: {:?}, raw response: {}",
+                parse_error,
+                response_text
+            );
+            return Err(error_stack::Report::from(OpenSearchError::ResponseError));
+        }
```
Comment (lsampras) [inline crates/analytics/src/search.rs:42]:
```suggestion
        er@Err(parse_error) => {
            tracing::error!(
                "Failed to parse response: {:?}, raw response: {}",
                parse_error,
                response_text
            );
            return er.change_context(OpenSearchError::ResponseError);
        }
```
can we do something like this ?

Comment (lsampras) [reply crates/analytics/src/search.rs:42]: Can we do it a bit more functional ?

something like

```rs
.text()
.change_context(OpensearchError::ResponseBodyError)
.and_then(|body|
    serde_json::from_str(&body)
    .change_context(OpensearchError::ResponseDeserializationError)
    .attach_printable(&body)
)?
```

**Review by lsampras — APPROVED** [review]

File: crates/api_models/src/analytics/search.rs [hunk 6]
```diff
@@ -48,19 +48,50 @@ pub struct GetSearchResponse {
...
+
+impl OpensearchError {
+    pub fn error_type(&self) -> &str {
+        &self.error.error_type
+    }
+
+    pub fn reason(&self) -> &str {
+        &self.error.reason
+    }
+}
```
[the remark this answers is not in the document]
Comment (tsdk02) [reply crates/api_models/src/analytics/search.rs:83]: Not being used, removed it.

File: crates/analytics/src/search.rs (hunk 5 above)
Comment (tsdk02) [reply crates/analytics/src/search.rs]: Done

**Review by lsampras — APPROVED** [review]

**Review by ivor-juspay — APPROVED** [review]

**Pull request merged**